Add Karthus Q

MingSaber976 opened the pull request:
use IGameObject and GetUnitsInRange to detect its damage.
(This time I use branch to PR it xD)

MingSaber976 committed:
Add Karthus Q

use IGameObject and GetUnitsInRange to detect its damage.

MingSaber976 committed:
Remove some useless lines

MingSaber976 committed:
Remove unused "using"

YasuoLtBlue reviewed (approved):
Works <3

FurkanS1821 commented on Champions/Karthus/Q.cs:
```diff
@@ -0,0 +1,72 @@
+            CreateTimer(0.5f, () =>
+            {
+                IGameObject m = AddMinion(owner, "TestCube", "TestCube", spell.X, spell.Y);
+                var range = GetUnitsInRange(m, 100, true);
```
AFAIK there should be a method which takes coordinates as arguments. If there is, using that is a better approach IMO.

FurkanS1821 commented on Champions/Karthus/Q.cs:
```diff
@@ -0,0 +1,72 @@
+        {
+            AddParticleTarget(owner, "Karthus_Base_Q_Hand_Glow.troy", owner, 1, "R_Hand");
+            AddParticle(owner, "Karthus_Base_Q_Point.troy", spell.X, spell.Y);
+            AddParticle(owner, "Karthus_Base_Q_Ring.troy", spell.X, spell.Y);
```
AFAIK this ring is shown in red to the opposing team, but I am not sure if that can even be implemented with current API. If there is a team-special particle thing added, use that please. If not, please open an issue about it, or leave a comment.

MingSaber976 replied:
currently we can't add ring from team
"SetToVisibleByTeam" wasn't do the thing, so I just use the Particle, at least it looks work.

MingSaber976 replied:
Still this is the problem part, but I can't deal with it because currently there is no function can fix that.
I tried `SetVisibleByTeam` but didn't work.

MingSaber976 replied:
So I gonna keep this codes because currently we can't make Particle invisible when you add that.

FurkanS1821 commented on Champions/Karthus/Q.cs:
```diff
@@ -0,0 +1,72 @@
+                
+                foreach (var units in range)
+                {
+                    if (units.Team != owner.Team && units is IChampion || units is IMinion)
```
`&&` has priority over `||`, so this will allow damaging allied minions. Put the last two checks into paranthesis to fix this.

MingSaber976 replied:
I will add `.Where(x => x.Team == CustomConvert.GetEnemyTeam(owner.Team))` after "range",
so it will only damage to enemy team

FurkanS1821 commented on Champions/Karthus/Q.cs:
```diff
@@ -0,0 +1,72 @@
+                var ap = owner.Stats.AbilityPower.Total;
+                var damage = 20f + spell.Level * 20f + ap * 0.3f;
+                
+                foreach (var units in range)
```
Yes, this looks cool, but for an outsider reading lines below, it is confusing. Please rename these variables for readability.

FurkanS1821 commented on Champions/Karthus/Q.cs:
```diff
@@ -0,0 +1,72 @@
+                foreach (var units in range)
+                {
+                    if (units.Team != owner.Team && units is IChampion || units is IMinion)
+                    {
```
This block seems to have a lot of wrong things.
1. Double damage is applied if it hits *one* unit, not *two*.
2. `range` variable includes things which may be allied and/or things you cannot hit (e.g. wards, turrets or untargetable things). Using its count is problematic.

MingSaber976 replied:
about that part, I'll take a look,because current code I used "AddMinion", so it will count to 2.
I will use Particle instead, so It will be 1.
count part I will figure out how to do that.

FurkanS1821 commented on Champions/Karthus/Q.cs:
```diff
@@ -0,0 +1,72 @@
+        {
+        }
+
+        public void OnStartCasting(IChampion owner, ISpell spell, IAttackableUnit target)
```
You spawn the mine and damage everyone *at the same frame*. Damage should be under `OnFinishCasting()`.

FurkanS1821 reviewed (dismissed):
Doesn't work *just well*, unlike what's stated above.

MingSaber976 committed:
Update Q.cs

fixing patch
1.renamed `range` to `spellPosition` so that will be more clear to read.
2.`IGameObject m` change type from `Minion` to `Particle`
3.add `.Where(x => x.Team == CustomConvert.GetEnemyTeam(owner.Team))` so that will only damage to enemy
------
bug: Particle can't be set to Visible by team.

danil179 commented on Champions/Karthus/Q.cs:
```diff
@@ -0,0 +1,72 @@
+            CreateTimer(0.5f, () =>
+            {
+                IGameObject m = AddParticle(owner, "Karthus_Base_Q_Explosion.troy", spell.X, spell.Y);
+                var spellPosition = GetUnitsInRange(m, 150, true);
```
This is NOT `spellPosition`. Rename it to `affectedUnits`.

danil179 commented on Champions/Karthus/Q.cs:
```diff
@@ -0,0 +1,72 @@
+                var ap = owner.Stats.AbilityPower.Total;
+                var damage = 20f + spell.Level * 20f + ap * 0.3f;
+                
+                foreach (var units in spellPosition
```
Should be `unit` instead of `units`.

danil179 commented on Champions/Karthus/Q.cs:
```diff
@@ -0,0 +1,72 @@
+                {
+                    if (units is IChampion || units is IMinion)
+                    {
+                        if (spellPosition.Count == 0)
```
Code will never enter this `if`. Put this code before/after the `foreach` loop.

MingSaber976 replied:
You mean `Count == 0` part?

MingSaber976 replied:
if it so, i'll changed codes.

danil179 commented on Champions/Karthus/Q.cs:
```diff
@@ -0,0 +1,72 @@
+            AddParticle(owner, "Karthus_Base_Q_Point.troy", spell.X, spell.Y);
+            AddParticle(owner, "Karthus_Base_Q_Ring.troy", spell.X, spell.Y);
+            AddParticle(owner, "Karthus_Base_Q_Skull_Child.troy", spell.X, spell.Y);
+            CreateTimer(0.5f, () =>
```
I'm not sure if that should be a timer at all. There are functions for `FinishCasting` and `ApplyEffects`. Please check if some of these functions will help you removing this timer.

MingSaber976 replied:
All fixed.

danil179 reviewed (changes requested):
Didn't test if that works, and relay on last comments.

Seems ok, but there are some problems. This is better than current state and therefore I think it should get merged after fixing the code style problems.

MingSaber976 committed:
Update Q.cs

1.Change `spellPosition` to `affectUnits`
2.change the name `units` to `unit`
3.make codes into `OnFinishChasting`
4.let `affectedUnits.Count == 0` part before the `foreach()`

danil179 commented on Champions/Karthus/Q.cs:
```diff
@@ -0,0 +1,68 @@
+        {
+            IGameObject m = AddParticle(owner, "Karthus_Base_Q_Explosion.troy", spell.X, spell.Y);
+                var affectedUnits = GetUnitsInRange(m, 150, true);
+                var ap = owner.Stats.AbilityPower.Total;
```
The indention is wrong.

MingSaber976 replied:
Done.

danil179 reviewed (changes requested):
Wrong indention.

MingSaber976 committed:
Update Q.cs

fixing indention.

MythicManiac commented:
@FurkanS1821 is this okay now?
